flink-runtime/src/main/java/org/apache/flink/runtime/util/ResourceCounter.java

```diff
@@ -35,11 +37,12 @@
  * associated counts. The counts are always positive (> 0).
  */
 public final class ResourceCounter {
```

Comment:
Why do you make this class mutable? It was designed to be immutable. I think there is good value in leaving this class immutable.

Reply:
I make it mutable mainly for performance concerns. If someone uses it like a "counter", `add` and `substract` resource profile for a lot of times, the construct of `ResourceCounter` might be expensive. It may also put a lot of pressure on the GC. For instance, every time a slot is allocated, we will construct 2 `ResourceCounter`s in `ResourceTracker#notifyAcquiredResource`.

Reply:
Each time we add a `ResourceProfile` to it, all the `ResourceProfile`s recorded will be put into a new map, with O(nlogn) complexity. For coarse-grained resource management, that's might be ok since we only have one type of requirement and one type of slot(for standalone mode, that assumption might be broken). However, for fine-grained resource management, performance can get bad.

Reply:
Have you run some experiments or are you suspecting bad performance @KarmaGYZ? If it is the latter case, then I would consider this a case of premature optimization.

Reply:
@tillrohrmann I write a test in which I add two types of resources profile with 8000 times respectively(8000 * 8000 wordcount case). I run the test with mutable and immutable `ResourceCounter` for 100 times. The total time cost by immutable `ResourceCounter` is 700ms while the time cost by mutable `ResourceCounter` is 200ms. I think that might  make sense to do that optimization. Actually, I'm not quite sure about the benefit of keeping it immutable. But if the benefit really matters, I also agree to keep it immutable, since it is might not be the bottleneck of the scheduling atm.

Reply:
I don't think that microbenchmarks are very conclusive tbh. They usually neglect the fact that certain operations are not on the hot path and others are. I am not saying to revert these changes since you've done it but this is a good example for premature optimization which tries to optimize something which is not clear to be a problem. Mutability is always a problem in the sense that passing these objects around and then mutating it will cause other components to see these changes. Therefore, it is much safer to make objects immutable.

Reply:
Thanks for the explanation. I now understand your point.
> passing these objects around and then mutating it will cause other components to see these changes.
Agreed. When the `ResourceCounter` used as an internal state, we need to ensure it will not be exposed outside. If a method has to return a `ResourceCounter` instance, it should realize this can be modified outside.

> Also you don't have to worry about concurrency and these kind of things.
Regarding the concern of concurrency modification, I'm not quite sure about this. If we make the `ResourceCounter` immutable, then we need to change its reference every time we add/substract `ResourceProfile` to it. This means we might need a lock to protect it and make it `volatile`. Just making it immutable may not solve the problem.

An additional idea inspired by your comment. Maybe the `ResourceCounter` should only be used as a `Counter` instead of a `Record`. When a method needs to return a set of resources, immutable `ResourceRequirement` might be more suitable. When someone want to count a set of resources, they need mutable `ResourceCounter` instead. WDYT?

Reply:
Regarding the performance concerns, I agree that this might be premature optimization. We need a deeper investigation to prove it hurts performance. TBH, I don't think it can hurt the performance in scheduling a lot. I thought it would be more beneficial to optimize performance rather than keeping it immutable when I prepared this PR. Thanks for your explanation.

Reply:
The example with concurrency is that you don't have to lock accesses to an immutable object. If a component changes references to this object, then of course it needs to make sure that this happens under a lock if multiple threads can access it. But this is a concern of the component using the `ResourceCounter` and not the other way around.

Your last idea is also what I had in mind with the immutable `ResourceCounter`. The `ResourceCounter` here would be what you called immutable `ResourceRequirement`. It just happens that the implementation of the `SlotPool` and `SlotManager` internally use the same class because it was easier. There wouldn't speak anything against using a different representation which, for example, could be mutable and thus give us better performance. This is completely an implementation detail of the components returning a `ResourceCounter` in their API.

Reply:
Thanks for the explanation. I now know the story of the `slotPool` version `ResourceCounter`. Let's keep it immutable in this PR. If we meet any performance issue or regression in the future, we might go back to it. As the Chinese New Year is coming, I'll update this PR next week :)

Reply:
Alright. Ping me once you have updated the PR. I wish you a happy Chinese New Year :-)